qt5: cross compile with a clean separation of build and host binaries

eryngion opened the pull request:
## Description of changes

By default when cross compiling QT builds all binaries that are marked in their .pro files with ```option(host_build)``` for the build platform instead of the host platform, and those binaries end up in the dev outputs (maybe not only there) of packages intended for the host platform. While there is some expectation of this situation in Nix [1], it prevents us from easily swapping natively and cross built dependencies, i.e. "lazy cross-compiling" and the host platform doesn't get the full set of binaries.

Here I'm attempting to have only host-targeted binaries in the cross-compiled packages and somehow live with this new situation. Even if this approach fails, I have here a couple of fixes that can helpful elsewhere.

I'll direct this PR to staging when we will be done discussing it.

[1] ```patchShebangs``` expects that binaries in dev outputs will be run on the build platform, and ```propagatedNativeBuildInputs```, being implemented as an in-store dependency, essentially adds the same hard dependency on the platform where the package was built.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux (cross)
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Artturin committed:
qt5.wrapQtAppsHook: move libs to depsTargetTargetPropagated

https://github.com/NixOS/nixpkgs/blob/master/doc/builders/special/makesetuphook.section.md

Commit:
distccMasquerade: add ar

Commit:
qt5.full: add disallowedReferences for qmake from buildHost

Commit:
qt5: fix discovery of QT modules in cross builds

Without it qtwebchannel will fail to find some library includes,
qtquickcontrols will refuse to build without properly failing the
build:
```
Some of the required modules (qtHaveModule(quick)) are not available.
Skipped.
```
And there are probably more problems, not always obvious.

Commit:
qt5.qtdeclarative: fix cross for qtdeclarative and some of it's users

If you're a nativeBuildInput somewhere, you probably shouldn't propagate
qtbase.

Commit:
qt5.qtwayland: fix cross compilation and dependencies

It needs only qtdeclarative to build, not qtquickcontrols.

Commit:
qt5.qttools: fix detection of lrelease and other tools in cross builds with qttools as a buildInput

And don't propagate qtbase to avoid pulling in a second qtbase.dev in
cross builds.

eryngion commented on pkgs/development/libraries/qt-5/modules/qtbase-bootstrap.nix:
This whole thing should be merged back to qtbase.nix once I'm done fiddling.

eryngion commented on pkgs/development/libraries/qt-5/hooks/qtbase-setup-hook.sh:
```diff
@@ -56,7 +52,7 @@ qmakePathHook() {
     fi
 }
-envBuildHostHooks+=(qmakePathHook)
+envHostTargetHooks+=(qmakePathHook)
```
Maybe qmakePathHook should be run for both types of packages? Not sure yet.

eryngion commented:
@amjoseph-nixpkgs @Artturin

eryngion commented on pkgs/development/libraries/qt-5/5.15/default.nix:
```diff
@@ -311,41 +333,19 @@ let
-        propagatedBuildInputs = [ qtbase.dev makeBinaryWrapper ]
+        propagatedBuildInputs = [ makeBinaryWrapper ];
+        depsTargetTargetPropagated = [ qtbase.dev ]
           ++ lib.optional stdenv.isLinux qtwayland.dev;
```
We have no packages in this repository that use `wrapQtAppsHook` and do not depend on `qtbase`. Should we really do this propagation? Is there a real benefit of having `qtwayland` paths in the wrapper if the package itself does not depend on it?

eryngion commented:
`goldendict` now builds and runs, which is a good sign.

eryngion commented on pkgs/development/libraries/qt-5/modules/qttools.nix:
```diff
@@ -1,9 +1,20 @@
 qtModule {
   pname = "qttools";
-  propagatedBuildInputs = [ qtbase qtdeclarative ];
-  outputs = [ "out" "dev" "bin" ];
```
Maybe I shouldn't change the order here?

Commit:
Revert "qt5.qtwebchannel: omit "bin" output when cross compiling"

Not sure why it was done.

This reverts commit 49e8d9db2f83003c0c5ac6c2df25104390f60519.

Commit:
qt5.qtwebengine: update cross compilation support

Commit:
python3Packages.pyqt5: update cross

amjoseph-nixpkgs commented:
> By default when cross compiling QT builds all binaries that are marked in their .pro files with `option(host_build)` for the build platform instead of the host platform, and those binaries end up in the dev outputs (maybe not only there) of packages intended for the host platform.

Hey I'm very in favor of this in principle, but it needs buy-in from the QT maintainers -- in other words, the native builds would need separation too.  There's already way too much divergence between the cross and native build paths.


> While there is some expectation of this situation in Nix [1], it prevents us from easily swapping natively and cross built dependencies

I don't think you can really expect to swap cross-built and native-built packages in nixpkgs.  It might work by accident but if so it's going to break randomly... it's not something that's expected to work and keep working.  In the long run we need `ca-derivations` in order to make that work.

eryngion commented:
> Hey I'm very in favor of this in principle

Good to hear this. I probably wouldn't have started this endeavor without your patches.

> it needs buy-in from the QT maintainers

Yeah, I'm aware of the general plans to get Nix packages closer to mainline QT way of doing things, and expect that there will be a discussion about these changes, and maybe not everything will get merged.

> the native builds would need separation too

It sounds like we have different understandings of separation here. The only change in the structure of user-facing binary packages/outputs that I needed so far is the separate `qtbase.qmake` output. It will be symlinked from `qtbase.dev` for backward compatibility [1]. And the need for that mostly arises from the self-imposed limitation of installing only one `qtbase.dev` at a time (which I still think could be a good thing).

At the build time we already partially rely on finding binaries in PATH ([`qtPrepareTool` stuff in `0003-qtbase-mkspecs.patch`](https://github.com/NixOS/nixpkgs/blob/ea2ee49a5af251ae6e3a15f545e3879cc7ac41f1/pkgs/development/libraries/qt-5/5.15/qtbase.patch.d/0003-qtbase-mkspecs.patch#L390-L404)), and for the most part I'm just introducing the same approach (a very Nixy alternative to the single `QT_HOST_BINS` directory) to a few more places. Which means that
a) it should mostly just work for the native builds (for now)
b) even without my changes a sizeable part of packages depending on, for example, `qtdeclarative` will fail to build natively the moment you set `strictDeps = true` for them. Including some of the modules of qt5. Which, I imagine, may be one of the scenarios you're expecting here.

> There's already way too much divergence between the cross and native build paths.

If we want to introduce `strictDeps` to native builds at some point in the future, then it's the native build path who should move somewhat closer to the cross. Although I certainly can imagine simpler ways of doing that :)

> I don't think you can really expect to swap cross-built and native-built packages in nixpkgs. It might work by accident but if so it's going to break randomly...

Well, in my limited experience with native deps in cross environment most packages actually "just work". QT seems to be a significant paint point in that regard that regard and in my opinion should support that usecase even if some other smaller packages don't do that. Some people seem to [go to great lengths](https://discourse.nixos.org/t/splitbuildinstall-split-buildphase-and-installphase-for-large-packages/15874) just to avoid building qtbase one more time :)

> In the long run we need `ca-derivations` in order to make that work.

So far I see aforementioned `propagatedNativeBuildInputs` and `patchShebangs` as major hurdles on Nix part here, the latter probably even being in conflict with `strictDeps` logic. But yeah, `ca-derivations` could play a role too.


[1] Yes, there are few packages out there that do things like `QMAKE=${qtbase.dev}/bin/qmake` and my changes will break the newly-acquired cross-compilation ability for them, if it works at all, but the native builds should be just fine. Seems like an easy fix in end-user packages.

Commit:
python3Packages.pyqtwebengine: update cross compilation support

I expect here that qtwebengine will propagate qtwebchannel and qtdeclarative and that
qtdeclarative will propagate itself as a nativeBuildInput. But maybe we
shouldn't rely on those propagations?

eryngion commented:
`pytq5` now builds "with everything" and the package is simpler.
`qwebengine` now builds, but for some reason without `qwebengine_convert_dict`. And something still insists on building the bundled ICU. (Maybe these things are connected.)
`pyqtwebengine` builds and the package is a bit simpler.
`qutebrowser` builds and runs ;)
`qtbase` + `cmake` combination now should "just work".

eryngion commented on pkgs/development/libraries/qt-5/modules/qtwebengine.nix:
```diff
@@ -30,46 +29,41 @@
+  isCrossBuild = stdenv.buildPlatform != stdenv.hostPlatform;
+
+  # qtwebengine requires its own very particular specially patched for qt version of gn
+  gnQtWebengine = with srcs.qtwebengine; buildPackages.gn.overrideAttrs {
```
If this thing is too scary, I can just remove it, but there will be one less binary shared between native and cross builds.

eryngion commented on pkgs/development/python-modules/pyqtwebengine/default.nix:
```diff
@@ -51,9 +47,6 @@ in buildPythonPackage (rec {
     qtwebengine
-  ] ++ lib.optionals (stdenv.buildPlatform != stdenv.hostPlatform) [
-    qtwebchannel
-    qtdeclarative
```
I expect here that `qtwebengine` will propagate `qtwebchannel` and `qtdeclarative` and that `qtdeclarative` will propagate itself as a `nativeBuildInput`. But maybe we shouldn't rely on those propagations?

eryngion commented:
@ofborg build qt5.full

Commit:
qt5.qtbase: add preliminary support for cross on Darwin

eryngion commented:
Cross on Darwin seems to be broken atm even for [the simplest of cases](https://logs.ofborg.org/?key=nixos/nixpkgs.267311&attempt_id=dbeb1e97-8c2f-4afa-acdf-0593e88c747f).

Commit:
qt5: simplify package functions' definitions and calls

Rely on callPackage whenever possible.

eryngion commented:
A compromise approach with less patches would be to build mixed-arch packages but still teach `qtbase` to use `qmake` from the `buildHost` version of itself, i.e. the `XQMAKESPEC=` stuff in `qtbase-setup-hook.sh`. But I don't think that it's a good packaging.

Commit:
qt5.qtbase: remove broken QMAKEMODULES code

It never worked and it's just an attempt to duplicate the logic
from qmake (i.e. QMAKEPATH -> QMAKEMODULES translation) that qmake
does all by itself.

Commit:
qt5.qtbase: fix mysql support for cross builds

Commit:
qt5: rework bootstrapping

Add to qtbase support for building a small bootstrap package.

Move qmake and other tools from qtbase.dev to qtbase.qmake
to avoid pulling in a second version of qtbase.dev in cross builds.

Use hidden bootstrap packages for qtbase and qttools instead of
overrideScope-based staging process to solve some spurious builds
and infinite recursions.

This commit by itself will allow to cross compile only qttranslations
and qtbase.

Commit:
qt5.qtbase: always build all packaged binaries for the hostPlatform

Also teach the setup hook, qmake and cmake scripts to search various
build tools in PATH and use them for cross compilation.

This will allow to cross compile packages that require only qtbase and
qmake.

Commit:
qt5.qtbase: remove the unused withQttranslation parameter

Commit:
qt5.qtbase: support substituting a cross-built qtbase dependency with a native-built one

For this we need to dynamically detect cross compilation in qtbase-setup-hook.sh
and in mkspecs, and stop unconditionally using binary lndir in propagatedNativeBuildInputs.

For now qtbase substitution will work only in one (most used) direction, and in many cases
will additionally require to manually add a buildHost version of qtbase.qmake to
nativeBuildInputs of the package that depends on qtbase.

eryngion commented:
@Izorkin can you check that your configuration with `environment.noXlibs = true` still builds correctly on this branch and show me how to reproduce the problem if it fails? Thanks.

Izorkin commented:
@eryngion error build murmur:
```
       … while calling 'dischargeProperties'

         at /home/user/works/src-nix/nixpkgs/lib/modules.nix:896:25:

          895|   */
          896|   dischargeProperties = def:
             |                         ^
          897|     if def._type or "" == "merge" then

       … while evaluating derivation 'murmur-1.4.287'
         whose name attribute is located at /home/user/works/src-nix/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:348:7

       … while evaluating attribute 'nativeBuildInputs' of derivation 'murmur-1.4.287'

         at /home/user/works/src-nix/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:392:7:

          391|       depsBuildBuild              = elemAt (elemAt dependencies 0) 0;
          392|       nativeBuildInputs           = elemAt (elemAt dependencies 0) 1;
             |       ^
          393|       depsBuildTarget             = elemAt (elemAt dependencies 0) 2;

       … while evaluating derivation 'wrap-qt5-apps-hook'
         whose name attribute is located at /home/user/works/src-nix/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:348:7

       … while evaluating attribute 'depsTargetTargetPropagated' of derivation 'wrap-qt5-apps-hook'

         at /home/user/works/src-nix/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:403:7:

          402|       propagatedBuildInputs       = elemAt (elemAt propagatedDependencies 1) 1;
          403|       depsTargetTargetPropagated  = elemAt (elemAt propagatedDependencies 2) 0;
             |       ^
          404|

       … while evaluating derivation 'qtbase-5.15.11'
         whose name attribute is located at /home/user/works/src-nix/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:348:7

       … while evaluating attribute 'configureFlags' of derivation 'qtbase-5.15.11'

         at /home/user/works/src-nix/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:406:7:

          405|       # This parameter is sometimes a string, sometimes null, and sometimes a list, yuck
          406|       configureFlags =
             |       ^
          407|         configureFlags

       … while evaluating derivation 'qttranslations-5.15.11'
         whose name attribute is located at /home/user/works/src-nix/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:348:7

       … while evaluating attribute 'disallowedReferences' of derivation 'qttranslations-5.15.11'

         at /home/user/works/src-nix/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:514:7:

          513|     optionalAttrs (attrs ? disallowedReferences) {
          514|       disallowedReferences =
             |       ^
          515|         map unsafeDerivationToUntrackedOutpath attrs.disallowedReferences;

       error: value is a set while a list was expected
```

Izorkin commented:
Update PR, next error:
```
error: cannot download cross_build_mysql.diff from any mirror
building '/nix/store/a4mxkza5v8rvxjc9446aj32cqxym0d47-crontabs.drv'...
Running phase: unpackPhase
unpacking source archive /nix/store/0jx5728jmsaza29mj1z0s5z82iacqcll-avahi-0.8.tar.gz
error: builder for '/nix/store/v80rzmgkhsn9mb403z99hpwc9f63i4m4-cross_build_mysql.diff.drv' failed with exit code 1;
       last 7 log lines:
       >
       > trying https://sources.debian.org/data/main/q/qtbase-opensource-src/5.15.11%2Bdfsg-1/debian/patches/cross_build_mysql.diff
       >   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
       >                                  Dload  Upload   Total   Spent    Left  Speed
       >   0   265    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
       > curl: (22) The requested URL returned error: 404
       > error: cannot download cross_build_mysql.diff from any mirror
       For full logs, run 'nix log /nix/store/v80rzmgkhsn9mb403z99hpwc9f63i4m4-cross_build_mysql.diff.drv'.

```

Izorkin commented:
Thanks, PR working with `environment.noXlibs = true`.

Artturin commented:
Rebased.

eryngion commented:
![detected](https://github.com/NixOS/nixpkgs/assets/13999967/025f8151-02b3-44e0-9cba-7efbabbc1d99)

Commit:
qt5.qtbase: strictDeps works without mysql in nativeBuildInputs

Commit:
qt5.qtwebkit: fix build and use hyphen and woff2 from pkgs

hyphen's sources in tarbals are identical, and the version from
pkgs properly supports cross compilation.

Commit:
qt5.qmake: make it usable again for packages that pass file names in qmakeFlags

but don't use __structuredAttrs

Commit:
python3Packages.pyqtwebengine: nixfmt

Commit:
Revert "qt5.qtwebengine: Pin to GCC 14"

This reverts commit 6e27bc915b374bfdc7cabe75018f16b6c623fa91.

Commit:
qt5.qtdeclarative: fix building with gcc15 and remove unused qtsvg dependency

eryngion commented:
Yes, I'm still interested and would appreciate your help. I'm building a fresh rebase rn and will push it probably tomorrow.

And I'm also still willing to look into cross compilation of Qt6 once this PR gets out of limbo one way or another. @K900 Is the end of Qt5 nigh? I somehow keep reverting and fixing up things here that land in master. Some of these commits do not depend on this PR and affect not only cross compilation. Would you be interested in a separate PR or PRs for them? I made a small attempt two years ago but it received little attention.

K900 commented:
Qt5 will probably be around for a while longer still, so I think this may be worth doing. Splitting the PR is generally a good call when possible. Also, Qt6 can already kinda sorta cross compile, so it's significantly less bad (but still bad, because cmake).

Commit:
qt5.qtwebengine: fix building with gcc 15, use more system librares, and
propagate less of them

doronbehar commented:
For reference:

- https://github.com/NixOS/nixpkgs/pull/487060
